### PR TITLE
Improve iA command for MIPS ##bin

### PR DESCRIPTION
**Pull request by junchao-loongson**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

Before MIPS32/64, a number of versions of the MIPS architecture were written down.
I added the  function that the iA command can output target ISA

I want to pass the ISA information to the gnu disassembler, do you have any suggestions?

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by XVilka** [comment]
You can use `asm.cpu` for that.

**Comment by junchao-loongson** [comment]
> You can use `asm.cpu` for that.

Thank you, but it seem that asm.cpu infomation is not available in the disassembly plugin.

R_API const char* r_config_get(RConfig *cfg, const char *name) 
It requires RConfig parameter

**Comment by junchao-loongson** [comment]
> You can use `asm.cpu` for that.
libr/asm/arch/mips/gnu/mips-dis.c
```
 546   /* If an ELF "newabi" binary, use the n32/(n)64 GPR names.  */
 547   if (info->flavour == bfd_target_elf_flavour && info->section != NULL)
 548     {
 549       Elf_Internal_Ehdr *header;
 550 
 551       header = elf_elfheader (info->section->owner);
 552       if (is_newabi (header)) {
 553           mips_gpr_names = mips_gpr_names_newabi;
 554       }
 555     }
```

I want the above code to work, but I don't know how to set info->section(disassemble_info->section).
I want to be able to automatically set ISA based on the header information of the elf file.
Do you have any suggestions or tips？

**Comment by XVilka** [comment]
@junchao-loongson you can see how it's done for TMS320* family - see `tms320_f_get_cpu()` and `tms320_f_set_cpu()`. See also `asm.features` and search for `features` in `libr/asm/`

**Comment by junchao-loongson** [comment]
> @junchao-loongson you can see how it's done for TMS320* family - see `tms320_f_get_cpu()` and `tms320_f_set_cpu()`. See also `asm.features` and search for `features` in `libr/asm/`
> 
> See also [libr/asm/p/asm_tms320.c](https://github.com/radareorg/radare2/blob/master/libr/asm/p/asm_tms320.c#L75):
> 
> ```c
> static int tms320_disassemble(RAsm *a, RAsmOp *op, const ut8 *buf, int len) {
> 	if (a->cpu && r_str_casecmp (a->cpu, "c54x") == 0) {
> 		tms320_f_set_cpu (&engine, TMS320_F_CPU_C54X);
> 	} else if (a->cpu && r_str_casecmp(a->cpu, "c55x+") == 0) {
> 		tms320_f_set_cpu (&engine, TMS320_F_CPU_C55X_PLUS);
> 	} else if (a->cpu && r_str_casecmp(a->cpu, "c55x") == 0) {
> 		tms320_f_set_cpu (&engine, TMS320_F_CPU_C55X);
> 	} else {
> #if CAPSTONE_HAS_TMS320C64X
> 		if (a->cpu && !r_str_casecmp (a->cpu, "c64x")) {
> 			return tms320c64x_disassemble (a, op, buf, len);
> 		}
> #endif
> 		r_asm_op_set_asm (op, "unknown asm.cpu");
> 		return op->size = -1;
> 	}
> 	op->size = tms320_dasm (&engine, buf, len);
> 	r_asm_op_set_asm (op, engine.syntax);
> 	return op->size;
> }
> ```

I understand now, thank you very much for your tips

File: libr/bin/format/elf/elf.c [hunk 1]
```diff
@@ -2102,6 +2102,28 @@ char* Elf_(r_bin_elf_get_arch)(ELFOBJ *bin) {
 	}
 }
 
+char* Elf_(r_bin_elf_get_head_flag)(ELFOBJ *bin){
+	if (bin->phdr && bin->ehdr.e_machine == EM_MIPS) {
+		const ut32 mipsType = bin->ehdr.e_flags & EF_MIPS_ARCH;
```
Comment (trufae) [inline libr/bin/format/elf/elf.c:2107]: spaces and tabs

**Review by trufae — CHANGES_REQUESTED** [review]
Please fix the indentation and syntax rules. but lgtm

File: libr/bin/format/elf/elf.c (hunk 1 above)
Comment (junchao-loongson) [reply libr/bin/format/elf/elf.c:2107]: Thank you for your patience. I will check the code format carefully next time.

File: libr/bin/bin.c [hunk 2]
```diff
@@ -1060,13 +1061,18 @@ R_API void r_bin_list_archs(RBin *bin, int mode) {
 			pj_ki (pj, "bits", bits);
 			pj_kn (pj, "offset", boffset);
 			pj_kn (pj, "size", obj_size);
+			if ( !strcmp(arch, "mips") ) {
```
Comment (trufae) [inline libr/bin/bin.c:1064]: the space goes before the (, not the !

File: libr/bin/bin.c [hunk 3]
```diff
@@ -1086,13 +1092,18 @@ R_API void r_bin_list_archs(RBin *bin, int mode) {
 				pj_ki (pj, "bits", bits);
 				pj_kn (pj, "offset", boffset);
 				pj_kn (pj, "size", obj_size);
+				if (!strcmp(arch, "mips")) {
```
Comment (trufae) [inline libr/bin/bin.c:1095]: missing space

File: libr/bin/bin.c [hunk 4]
```diff
@@ -1086,13 +1092,18 @@ R_API void r_bin_list_archs(RBin *bin, int mode) {
...
+					pj_ks (pj, "isa", h_flag);
+				}
 				if (machine) {
 					pj_ks (pj, "machine", machine);
 				}
 				pj_end (pj);
 				break;
 			default:
-				r_table_add_rowf (table, "nsnss", i, sdb_fmt ("0x%08" PFMT64x , boffset), obj_size, sdb_fmt("%s_%i", arch, bits), "");
+				str_fmt = strcmp(h_flag, "unknown_flag")? sdb_fmt ("%s_%i %s", arch, bits, h_flag) \
```
Comment (trufae) [inline libr/bin/bin.c:1104]: missing space

File: libr/bin/format/elf/elf.c [hunk 5]
```diff
@@ -2102,6 +2102,28 @@ char* Elf_(r_bin_elf_get_arch)(ELFOBJ *bin) {
 	}
 }
 
+char* Elf_(r_bin_elf_get_head_flag)(ELFOBJ *bin) {
+	if (bin->phdr && bin->ehdr.e_machine == EM_MIPS) {
+		const ut32 mipsType = bin->ehdr.e_flags & EF_MIPS_ARCH;
+		switch (mipsType) {
```
Comment (trufae) [inline libr/bin/format/elf/elf.c:2108]: still wrong indent

File: libr/bin/format/elf/elf.c [hunk 6]
```diff
@@ -2102,6 +2102,28 @@ char* Elf_(r_bin_elf_get_arch)(ELFOBJ *bin) {
...
+	    case EF_MIPS_ARCH_1:		return strdup ("mips1");
+	    case EF_MIPS_ARCH_2:		return strdup ("mips2");
+        case EF_MIPS_ARCH_3:		return strdup ("mips3");
+        case EF_MIPS_ARCH_4:		return strdup ("mips4");
+        case EF_MIPS_ARCH_5:		return strdup ("mips5");
+        case EF_MIPS_ARCH_32:		return strdup ("mips32");
```
Comment (trufae) [inline libr/bin/format/elf/elf.c:2114]: dont use tabs if its not the begining of the line

File: test/db/cmd/cmd_i [hunk 7]
```diff
@@ -337,6 +337,34 @@ va       true
...
+iAj
+EOF
+EXPECT=<<EOF
+0   0x00000000 1039368 mips_64 mips64r2 
+
+{"bins":[{"arch":"mips","bits":64,"offset":0,"size":1039368,"isa":"mips64r2","machine":"MIPS R3000"}]}
+EOF
+RUN
+
+
```
Comment (trufae) [inline test/db/cmd/cmd_i:366]: two unnecessary empty lines

**Comment by junchao-loongson** [comment]
> @junchao-loongson you can see how it's done for TMS320* family - see `tms320_f_get_cpu()` and `tms320_f_set_cpu()`. See also `asm.features` and search for `features` in `libr/asm/`
> 
> See also [libr/asm/p/asm_tms320.c](https://github.com/radareorg/radare2/blob/master/libr/asm/p/asm_tms320.c#L75):
> 
> ```c
> static int tms320_disassemble(RAsm *a, RAsmOp *op, const ut8 *buf, int len) {
> 	if (a->cpu && r_str_casecmp (a->cpu, "c54x") == 0) {
> 		tms320_f_set_cpu (&engine, TMS320_F_CPU_C54X);
> 	} else if (a->cpu && r_str_casecmp(a->cpu, "c55x+") == 0) {
> 		tms320_f_set_cpu (&engine, TMS320_F_CPU_C55X_PLUS);
> 	} else if (a->cpu && r_str_casecmp(a->cpu, "c55x") == 0) {
> 		tms320_f_set_cpu (&engine, TMS320_F_CPU_C55X);
> 	} else {
> #if CAPSTONE_HAS_TMS320C64X
> 		if (a->cpu && !r_str_casecmp (a->cpu, "c64x")) {
> 			return tms320c64x_disassemble (a, op, buf, len);
> 		}
> #endif
> 		r_asm_op_set_asm (op, "unknown asm.cpu");
> 		return op->size = -1;
> 	}
> 	op->size = tms320_dasm (&engine, buf, len);
> 	r_asm_op_set_asm (op, engine.syntax);
> 	return op->size;
> }
> ```

hello~  XVilka

I want to know how to set **disasm_obj.section**  here according to r_asm_t *a or r_asm_op_t *op(Or according to other parameters)
I want to set ABI and ISA based on ELF file header information

libr/asm/p/asm_mips_gnu.c 
`static int disassemble(struct r_asm_t *a, struct r_asm_op_t *op, const ut8 *buf, int len) {
	static struct disassemble_info disasm_obj;
	if (len < 4) {
		return -1;
	}
	buf_global = &op->buf_asm;
	Offset = a->pc;
	memcpy (bytes, buf, 4); // TODO handle thumb

	/* prepare disassembler */
	memset (&disasm_obj,'\0', sizeof (struct disassemble_info));
	mips_mode = a->bits;
	disasm_obj.arch = CPU_LOONGSON_2F;
	disasm_obj.buffer = bytes;`

File: libr/bin/format/elf/elf.c [hunk 8]
```diff
@@ -2106,24 +2106,22 @@ char* Elf_(r_bin_elf_get_head_flag)(ELFOBJ *bin) {
...
-        case EF_MIPS_ARCH_5:		return strdup ("mips5");
-        case EF_MIPS_ARCH_32:		return strdup ("mips32");
-        case EF_MIPS_ARCH_64:		return strdup ("mips64");
-        case EF_MIPS_ARCH_32R2:		return strdup ("mips32r2");
-        case EF_MIPS_ARCH_64R2:		return strdup ("mips64r2");
-        default :					return strdup (" Unknown mips ISA");
+	    case EF_MIPS_ARCH_1:        return strdup ("mips1");
+	    case EF_MIPS_ARCH_2:        return strdup ("mips2");
+        case EF_MIPS_ARCH_3:        return strdup ("mips3");
+        case EF_MIPS_ARCH_4:        return strdup ("mips4");
```
Comment (trufae) [inline libr/bin/format/elf/elf.c]: indent the 'case with tabs, the spaces after the `:` are optional, but keep them for now. also if you want to aviod 9999 calls to strdup you can just set a `const char *` and just return (retstr); at the end once

**Review by trufae — CHANGES_REQUESTED** [review]

File: libr/bin/format/elf/elf.c (hunk 8 above)
Comment (junchao-loongson) [reply libr/bin/format/elf/elf.c]: You seem to have marked outdated code

**Review by trufae — APPROVED** [review]

**Review by XVilka — APPROVED** [review]

**Pull request merged**